### PR TITLE
Properly truncate referenced src

### PR DESCRIPTION
**Pull request by liangdzou**
### Your checklist for this pull request

Please review the [guidelines for contributing](http://solidity.readthedocs.io/en/latest/contributing.html) to this repository.

Please also note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.

### Checklist
- [x] Code compiles correctly
- [x] All tests are passing
- [x] New tests have been created which fail without the change (if possible)
- [x] Used meaningful commit messages

### Description

This PR fixes the issue #4741 by trying to give more information for too long lines.
Basically, the solution gives at most 35 chars before and after the string we want to print if the line is too long. The previous solution simply prints the string out which is lack of the context of the problem we want to show to a user.

This PR is tested with 4 possible situations, i.e. no side, left side, right side, or both sides are shorter than 35 chars. Two more edge tests are also added to test that it prints "..." when the corresponding edge has more than 35 chars.

### CODE REVIEW

File: libsolidity/interface/SourceReferenceFormatter.cpp [hunk 1]
```diff
@@ -55,8 +55,11 @@ void SourceReferenceFormatter::printSourceLocation(SourceLocation const* _locati
 		}
 		if (line.length() > 150)
 		{
-			line = " ... " + line.substr(startColumn, locationLength) + " ... ";
-			startColumn = 5;
+			int len = line.length();
+			line = line.substr(max(0, startColumn - 35), min(startColumn, 35) + min(locationLength + 35, len - startColumn));
+			line = startColumn > 35 ? " ... " + line : line;
```
Comment (chriseth) [inline libsolidity/interface/SourceReferenceFormatter.cpp]: I think I would prefer `if (startColumn > 35) { line = "... " + line; }`, then you could also pull up the conditional adjustment of `startColumn` below.

Also, why do you use `" ... "` here, but `" ..."` below?

File: test/cmdlineTests.sh [hunk 2]
```diff
@@ -156,6 +156,18 @@ printTask "Testing passing empty remappings..."
 test_solc_file_input_failures "${0}" "=/some/remapping/target" "" "Invalid remapping: \"=/some/remapping/target\"."
 test_solc_file_input_failures "${0}" "ctx:=/some/remapping/target" "" "Invalid remapping: \"ctx:=/some/remapping/target\"."
 
+printTask "Testing passing location printing..."
+(
+cd "$REPO_ROOT"/test/errorReports/
```
Comment (chriseth) [inline test/cmdlineTests.sh]: `errorReports` is a little unspecific inside the `test` directory. Perhaps `cmdlineErrorReports`? We could also move the whole cmdline test suite into its own subdirectory.

File: test/errorReports/too_long_line.sol.log [hunk 3]
```diff
@@ -0,0 +1,7 @@
+Warning: This is a pre-release compiler version, please do not use it in production.
```
Comment (chriseth) [inline test/errorReports/too_long_line.sol.log]: You might have to find a way to filter this out - it will not be there for a release version and thus the tests would fail for the release version...

File: test/errorReports/too_long_line_both_sides_short.sol.log [hunk 4]
```diff
@@ -0,0 +1,7 @@
+Warning: This is a pre-release compiler version, please do not use it in production.
+too_long_line_both_sides_short.sol:1:1: Warning: Source file does not specify required compiler version!
+contract C {
+^ (Relevant source part starts here and spans across multiple lines).
+too_long_line_both_sides_short.sol:2:15: Error: Identifier not found or not unique.
+   function f(announcementTypeXXXXXXXXXXXXXXXXXXX ... XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX Type,
+              ^-------------------------------------------------------------------------^
```
Comment (chriseth) [inline test/errorReports/too_long_line_both_sides_short.sol.log]: I think this is probably out of scope for this change, but it would be great to have a matching 
` ^--------------------------------- ... ------------------------------------^` here

File: libsolidity/interface/SourceReferenceFormatter.cpp (hunk 1 above)
Comment (liangdzou) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: My first implementation uses both " ... ". However, our current test system will detect any file for tailing whitespaces. So, I have to modify it to avoid the reference file be checked.
I agree. This is ugly.

File: test/errorReports/too_long_line.sol.log (hunk 3 above)
Comment (liangdzou) [reply test/errorReports/too_long_line.sol.log]: sure, thx.

File: libsolidity/interface/SourceReferenceFormatter.cpp (hunk 1 above)
Comment (liangdzou) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: I will use " ... " for both cases and try to avoid the test problem.

Comment (chriseth) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: I would actually argue that you do not need a ` ` after the `...` if nothing else follows.

Comment (liangdzou) [reply libsolidity/interface/SourceReferenceFormatter.cpp]: I think if there is nothing follows, there will not print the " ... ", see the test case too_long_line_right_short.sol and its reference.

File: test/cmdlineTests.sh (hunk 2 above)
Comment (liangdzou) [reply test/cmdlineTests.sh]: sure

**Comment by liangdzou** [comment]
If there is nothing follows, we will not print the " ... ", see the test case too_long_line_right_short.sol and its reference. Also, boundary tests are also included to test the cases have 35 and 36 chars before and after the "thing" we want to print.

**Comment by liangdzou** [comment]
@chriseth if you mean the case "too_long_line_edge_out.sol.ref", this case is intentionally designed to have 36 chars on the right-hand side. So, there is one "{" replaced by " ... ".

File: scripts/detect_trailing_whitespace.sh [hunk 5]
```diff
@@ -4,7 +4,7 @@ REPO_ROOT="$(dirname "$0")"/..
 
 (
 cd $REPO_ROOT
-WHITESPACE=$(git grep -n -I -E "^.*[[:space:]]+$" | grep -v "test/libsolidity/ASTJSON\|test/compilationTests/zeppelin/LICENSE")
+WHITESPACE=$(git grep -n -I -E "^.*[[:space:]]+$" | grep -v "test/libsolidity/ASTJSON\|test/compilationTests/zeppelin/LICENSE" | egrep -v "test/.*/.*\.ref")
```
Comment (ekpyron) [inline scripts/detect_trailing_whitespace.sh]: Why do you exclude trailing whitespace here? You should instead fix the code, so that ``solc`` does not output trailing whitespace, so that you don't need them in the test references.

**Review by ekpyron — CHANGES_REQUESTED** [review]

File: scripts/detect_trailing_whitespace.sh (hunk 5 above)
Comment (liangdzou) [reply scripts/detect_trailing_whitespace.sh]: My fix brings trailing whitespace since I will add " ... " at the end of a line. I will try to avoid it.

Comment (ekpyron) [reply scripts/detect_trailing_whitespace.sh]: Yes, if it's not too much effort, it would be nice not to avoid trailing whitespace in the output.

Comment (liangdzou) [reply scripts/detect_trailing_whitespace.sh]: It is done. Uploaded. :-)

**Review by ekpyron — DISMISSED** [review]

**Comment by ekpyron** [comment]
Looks fine to me now - for the edge cases one could have considered to only replace with `` ...``, if `` ...`` is in fact shorter than without the replacement (i.e. not replace the ``{`` in the edge case, but only replace, if it's more than four characters longer than the threshold), but I'm fine with it as it is as well.

**Comment by chriseth** [comment]
Could you please squash this into a single commit? Then it can be merged. Thanks a lot!

**Comment by liangdzou** [comment]
> Could you please squash this into a single commit? Then it can be merged. Thanks a lot!

@chriseth it is now squashed into one commit :-)

**Comment by liangdzou** [comment]
@ekpyron I have squashed the commit together according to @chriseth 's requirement. :-)

**Review by chriseth — APPROVED** [review]

**Comment by axic** [comment]
Could these not be covered by syntax tests?

**Comment by liangdzou** [comment]
@axic I added at least 8 cases for this PR

**Comment by axic** [comment]
I mean by the syntax tests harness and not via the newly added CLI tests.

**Comment by chriseth** [comment]
@axic no, the syntax tests only check the message itself, not the source reference.
